File: src/lib/tauri/credentials.ts
import { invoke } from "@tauri-apps/api/core";
import { isMockMode, isMockClaudeMode } from "./core";

// ── Credential / config status ────────────────────────────────────────────────

export interface CredentialStatus {
  anthropicApiKey: boolean;
  geminiApiKey: boolean;
  /** True when the user has enabled GitHub Copilot CLI delegation
   *  (`copilot_auth_method=copilot_cli`). Copilot has no API-key path. */
  copilotCli: boolean;
  /** True when the user has enabled Codex CLI delegation
   *  (`codex_auth_method=codex_cli`). Codex is CLI-only — used by
   *  the Commander panel's `codexAcp` backend. */
  codexCli: boolean;
  localLlmUrl: boolean;
  jiraBaseUrl: boolean;
  jiraEmail: boolean;
  jiraApiToken: boolean;
  jiraBoardId: boolean;
  bitbucketWorkspace: boolean;
  bitbucketEmail: boolean;
  bitbucketAccessToken: boolean;
  bitbucketRepoSlug: boolean;
}

export function credentialStatusComplete(s: CredentialStatus) {
  return (
    s.jiraBaseUrl &&
    s.jiraEmail &&
    s.jiraApiToken &&
    s.jiraBoardId &&
    s.bitbucketWorkspace &&
    s.bitbucketEmail &&
    s.bitbucketAccessToken &&
    s.bitbucketRepoSlug
  );
}

export function anthropicComplete(s: CredentialStatus) {
  return s.anthropicApiKey;
}

export function copilotComplete(s: CredentialStatus) {
  return s.copilotCli;
}

export function codexComplete(s: CredentialStatus) {
  return s.codexCli;
}

/** True when at least one AI provider (Anthropic, Gemini, Copilot, or local LLM) is configured. */
export function aiProviderComplete(s: CredentialStatus) {
  return s.anthropicApiKey || s.geminiApiKey || s.copilotCli || s.localLlmUrl;
}

/** All three auth credentials are present (board ID not required). */
export function jiraCredentialsSet(s: CredentialStatus) {
  return s.jiraBaseUrl && s.jiraEmail && s.jiraApiToken;
}

/** All three auth credentials are present (repo slug not required). */
export function bitbucketCredentialsSet(s: CredentialStatus) {
  return s.bitbucketWorkspace && s.bitbucketEmail && s.bitbucketAccessToken;
}

/** Fully ready: credentials + board ID configured. */
export function jiraComplete(s: CredentialStatus) {
  return s.jiraBaseUrl && s.jiraEmail && s.jiraApiToken && s.jiraBoardId;
}

/** Fully ready: credentials + repo slug configured. */
export function bitbucketComplete(s: CredentialStatus) {
  return (
    s.bitbucketWorkspace &&
    s.bitbucketEmail &&
    s.bitbucketAccessToken &&
    s.bitbucketRepoSlug
  );
}

// ── Credential commands ───────────────────────────────────────────────────────

/** All-false default used as the starting status when the underlying
 *  Tauri invoke fails — happens only when the app is loaded from the
 *  vite dev URL in a plain browser (e.g. chrome-devtools UI inspection).
 *  Mock overlays below fill in whichever surfaces the user has opted
 *  into. In a real Tauri runtime the invoke always succeeds, so this
 *  fallback is never reached. */
const EMPTY_STATUS: CredentialStatus = {
  anthropicApiKey: false,
  geminiApiKey: false,
  copilotCli: false,
  codexCli: false,
  localLlmUrl: false,
  jiraBaseUrl: false,
  jiraEmail: false,
  jiraApiToken: false,
  jiraBoardId: false,
  bitbucketWorkspace: false,
  bitbucketEmail: false,
  bitbucketAccessToken: false,
  bitbucketRepoSlug: false,
};

export async function getCredentialStatus(): Promise<CredentialStatus> {
  let status: CredentialStatus;
  try {
    status = await invoke<CredentialStatus>("credential_status");
  } catch {
    status = EMPTY_STATUS;
  }
  let merged: CredentialStatus = { ...status };
  if (isMockMode()) {
    merged = {
      ...merged,
      jiraBaseUrl: true,
      jiraEmail: true,
      jiraApiToken: true,
      jiraBoardId: true,
      bitbucketWorkspace: true,
      bitbucketEmail: true,
      bitbucketAccessToken: true,
      bitbucketRepoSlug: true,
    };
  }
  if (isMockClaudeMode()) {
    merged = {
      ...merged,
      anthropicApiKey: true,
    };
  }
  return merged;
}

export async function saveCredential(
  key: string,
  value: string,
): Promise<void> {
  return invoke("save_credential", { key, value });
}

export async function deleteCredential(key: string): Promise<void> {
  return invoke("delete_credential", { key });
}

/** Returns non-secret stored config values (URLs, email, workspace slug) for UI display. */
export async function getNonSecretConfig(): Promise<Record<string, string>> {
  return invoke<Record<string, string>>("get_non_secret_config");
}

File: src-tauri/src/commands/credentials.rs
use crate::storage::credentials;
use serde::Serialize;
use std::collections::HashMap;

// ── Allowed keys ───────────────────────────────────────────────────────────────

const ALLOWED_KEYS: &[&str] = &[
    "anthropic_api_key",
    "claude_auth_method",
    "gemini_api_key",
    "gemini_auth_method",
    "copilot_auth_method",
    "local_llm_url",
    "local_llm_api_key",
    "jira_base_url",
    "jira_email",
    "jira_api_token",
    "jira_account_id",
    "bitbucket_workspace",
    "bitbucket_email",
    "bitbucket_access_token",
    "bitbucket_username",
];

/// Keys whose values may be returned to the frontend (not secrets).
///
/// Includes the dual-store config keys (board id, repo slug, worktree paths,
/// base branch, terminal app) so the Settings → Configuration card can hydrate
/// them even when a legacy install left them in the credential store rather
/// than in `preferences.json`. `get_non_secret_config` lets `preferences`
/// override these on collision so the new storage location wins as soon as
/// the user re-saves the value.
const NON_SECRET_KEYS: &[&str] = &[
    "claude_auth_method",
    "gemini_auth_method",
    "copilot_auth_method",
    "local_llm_url",
    "jira_base_url",
    "jira_email",
    "jira_account_id",
    "jira_board_id",
    "bitbucket_workspace",
    "bitbucket_email",
    "bitbucket_username",
    "bitbucket_repo_slug",
    "repo_worktree_path",
    "repo_base_branch",
    "pr_review_worktree_path",
    "grooming_worktree_path",
    "pr_review_terminal",
];

// ── Tauri commands ─────────────────────────────────────────────────────────────

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CredentialStatus {
    pub anthropic_api_key: bool,
    pub gemini_api_key: bool,
    /// True when the user has enabled GitHub Copilot CLI delegation
    /// (`copilot_auth_method=copilot_cli`). Copilot has no API-key
    /// path — this flag is what "Copilot is configured" means.
    pub copilot_cli: bool,
    /// True when the user has enabled Codex CLI delegation
    /// (`codex_auth_method=codex_cli`). Codex is CLI-only — used by
    /// the Commander panel's `codexAcp` backend.
    pub codex_cli: bool,
    pub local_llm_url: bool,
    pub jira_base_url: bool,
    pub jira_email: bool,
    pub jira_api_token: bool,
    pub jira_board_id: bool,
    pub bitbucket_workspace: bool,
    pub bitbucket_email: bool,
    pub bitbucket_access_token: bool,
    pub bitbucket_repo_slug: bool,
}

impl CredentialStatus {
    pub fn anthropic_complete(&self) -> bool {
        self.anthropic_api_key
    }
    pub fn gemini_complete(&self) -> bool {
        self.gemini_api_key
    }
    pub fn copilot_complete(&self) -> bool {
        self.copilot_cli
    }
    pub fn codex_complete(&self) -> bool {
        self.codex_cli
    }
    pub fn local_llm_complete(&self) -> bool {
        self.local_llm_url
    }
    pub fn jira_complete(&self) -> bool {
        self.jira_base_url && self.jira_email && self.jira_api_token && self.jira_board_id
    }
    pub fn bitbucket_complete(&self) -> bool {
        self.bitbucket_workspace
            && self.bitbucket_email
            && self.bitbucket_access_token
            && self.bitbucket_repo_slug
    }
    pub fn all_complete(&self) -> bool {
        self.jira_complete() && self.bitbucket_complete()
    }
}

#[tauri::command]
pub fn credential_status() -> Result<CredentialStatus, String> {
    use crate::storage::preferences::get_pref;
    let has = |k: &str| credentials::cred_get(k).is_some();
    let has_config = |k: &str| get_pref(k).is_some() || credentials::cred_get(k).is_some();
    let copilot_method = credentials::cred_get("copilot_auth_method")
        .map(|m| m.trim().to_string())
        .unwrap_or_default();
    let codex_method = credentials::cred_get("codex_auth_method")
        .map(|m| m.trim().to_string())
        .unwrap_or_default();
    Ok(CredentialStatus {
        anthropic_api_key: has("anthropic_api_key"),
        gemini_api_key: has("gemini_api_key"),
        copilot_cli: copilot_method == "copilot_cli",
        codex_cli: codex_method == "codex_cli",
        local_llm_url: has("local_llm_url"),
        jira_base_url: has("jira_base_url"),
        jira_email: has("jira_email"),
        jira_api_token: has("jira_api_token"),
        jira_board_id: has_config("jira_board_id"),
        bitbucket_workspace: has("bitbucket_workspace"),
        bitbucket_email: has("bitbucket_email"),
        bitbucket_access_token: has("bitbucket_access_token"),
        bitbucket_repo_slug: has_config("bitbucket_repo_slug"),
    })
}

/// Return only non-secret stored values so the UI can pre-populate display fields.
/// Secret keys (API keys, tokens, passwords) are never included.
/// This merges non-secret keys from the credential store with all preference keys.
#[tauri::command]
pub fn get_non_secret_config() -> Result<HashMap<String, String>, String> {
    use super::preferences::get_preferences;
    let cred_map = credentials::load_map();
    let mut out: HashMap<String, String> = NON_SECRET_KEYS
        .iter()
        .filter_map(|&k| {
            cred_map
                .get(k)
                .filter(|v| !v.trim().is_empty())
                .map(|v| (k.to_string(), v.clone()))
        })
        .collect();

    if let Ok(prefs) = get_preferences() {
        out.extend(prefs);
    }

    Ok(out)
}

#[tauri::command]
pub fn save_credential(key: String, value: String) -> Result<(), String> {
    if !ALLOWED_KEYS.contains(&key.as_str()) {
        return Err("Unknown credential key.".to_string());
    }
    credentials::cred_set(&key, &value)
}

#[tauri::command]
pub fn delete_credential(key: String) -> Result<(), String> {
    if !ALLOWED_KEYS.contains(&key.as_str()) {
        return Err("Unknown credential key.".to_string());
    }
    credentials::cred_delete(&key)
}
